File: .ci/exact_rational_oracle_v150.py
from fractions import Fraction as Q


def crelu(x):
    return max(Q(0), x), max(Q(0), -x)


def sparsemax(xs):
    u = sorted(xs, reverse=True)
    tau = Q(0)
    for k in range(1, len(xs) + 1):
        candidate = (sum(u[:k]) - Q(1)) / Q(k)
        if k == len(xs) or u[k] <= candidate:
            tau = candidate
            break
    return [max(Q(0), x - tau) for x in xs], tau


def row_l1(row):
    return sum((abs(x) for x in row), Q(0))


def weight_l1(matrix):
    return sum((row_l1(row) for row in matrix), Q(0))


def path1(w1, w2):
    return sum(
        (abs(w2[o][h]) * row_l1(w1[h])
         for o in range(len(w2))
         for h in range(len(w1))),
        Q(0),
    )


def fmt(q):
    q = Q(q)
    return f"{q.numerator}/{q.denominator}"


def main():
    x = Q(-7, 3)
    pos, neg = crelu(x)
    assert pos - neg == x
    assert pos + neg == abs(x)

    scores = [Q(5, 4), Q(3, 4), Q(1, 2)]
    weights, tau = sparsemax(scores)
    assert weights == [Q(3, 4), Q(1, 4), Q(0)]
    assert sum(weights, Q(0)) == Q(1)
    assert all(w >= 0 for w in weights)
    for s, w in zip(scores, weights):
        if w == 0:
            assert s <= tau
        else:
            assert w == s - tau

    w1 = [[Q(1, 2), Q(-1, 3)], [Q(1, 4), Q(1, 5)]]
    w2 = [[Q(2, 3), Q(-3, 4)]]
    l1 = weight_l1(w1)
    p1 = path1(w1, w2)
    assert l1 == Q(77, 60)
    assert p1 == Q(77, 72)

    degrees = [Q(1)]
    for _ in range(4):
        degrees.append(degrees[-1] * 3)
    assert [int(d) for d in degrees] == [1, 3, 9, 27, 81]

    lines = [
        "oracle=python-fraction",
        f"crelu.reconstruct={fmt(pos - neg)}",
        f"crelu.abs={fmt(pos + neg)}",
        f"tsallis.tau={fmt(tau)}",
        "tsallis.weights=" + ",".join(fmt(w) for w in weights),
        f"weight_l1={fmt(l1)}",
        f"path1={fmt(p1)}",
        "degree_sequence=1,3,9,27,81",
        "status=PASS",
    ]
    print("\n".join(lines))


if __name__ == "__main__":
    main()

File: .ci/exact_rational_oracle_v150.hs
import Data.Ratio (Rational, (%), denominator, numerator)
import Data.List (sortBy)

q0 :: Rational
q0 = 0

crelu :: Rational -> (Rational, Rational)
crelu x = (max q0 x, max q0 (-x))

sparsemax :: [Rational] -> ([Rational], Rational)
sparsemax xs = (map (\x -> max q0 (x - tau)) xs, tau)
  where
    u = sortBy (flip compare) xs
    candidates = [ (sum (take k u) - 1) / fromIntegral k | k <- [1 .. length xs] ]
    admissible k t = k == length xs || (u !! k) <= t
    firstK = head [ (k, t) | (k, t) <- zip [1 .. length xs] candidates, admissible k t ]
    tau = snd firstK

rowL1 :: [Rational] -> Rational
rowL1 = sum . map abs

weightL1 :: [[Rational]] -> Rational
weightL1 = sum . map rowL1

path1 :: [[Rational]] -> [[Rational]] -> Rational
path1 w1 w2 = sum
  [ abs (w2 !! o !! h) * rowL1 (w1 !! h)
  | o <- [0 .. length w2 - 1]
  , h <- [0 .. length w1 - 1]
  ]

showR :: Rational -> String
showR q = show (numerator q) ++ "/" ++ show (denominator q)

main :: IO ()
main = do
  let x = (-7) % 3
      (pos, neg) = crelu x
      scores = [5 % 4, 3 % 4, 1 % 2]
      (weights, tau) = sparsemax scores
      w1 = [[1 % 2, (-1) % 3], [1 % 4, 1 % 5]]
      w2 = [[2 % 3, (-3) % 4]]
      l1 = weightL1 w1
      p1 = path1 w1 w2
      degrees = take 5 (iterate (*3) 1)
  if pos - neg /= x then error "CReLU reconstruction failed" else pure ()
  if pos + neg /= abs x then error "CReLU absolute decomposition failed" else pure ()
  if weights /= [3 % 4, 1 % 4, 0] then error "Tsallis-2 weights failed" else pure ()
  if sum weights /= 1 then error "Tsallis-2 normalization failed" else pure ()
  if l1 /= 77 % 60 then error "L1 weight norm failed" else pure ()
  if p1 /= 643 % 720 then error "1-path norm failed" else pure ()
  if degrees /= [1,3,9,27,81] then error "degree recurrence failed" else pure ()
  putStrLn "oracle=haskell-rational"
  putStrLn ("crelu.reconstruct=" ++ showR (pos - neg))
  putStrLn ("crelu.abs=" ++ showR (pos + neg))
  putStrLn ("tsallis.tau=" ++ showR tau)
  putStrLn ("tsallis.weights=" ++ comma weights)
  putStrLn ("weight_l1=" ++ showR l1)
  putStrLn ("path1=" ++ showR p1)
  putStrLn "degree_sequence=1,3,9,27,81"
  putStrLn "status=PASS"
  where
    comma [] = ""
    comma [a] = showR a
    comma (a:as) = showR a ++ "," ++ comma as
